### ISSUE TITLE
objetive-c support in protocols using BrightFutures

### ISSUE THREAD

**Issue by pellekrogholt** [Thomvis/BrightFutures#47]
I'm on version 1.0.1

I have a protocol for some couchbase lite utilities

```
import Foundation
import BrightFutures

@objc protocol CouchUtilsProtocol {
    var manager: CBLManager { get }

    func indexUpdate(
        databaseName: String,
        designDocName: String,
        viewName: String
    ) -> Future<Bool>

}
```

I have to add `@objc` because I use dependency injection ('Typhoon', '3.1.7') on compilation I get:

```
Method cannot be marked @objc because its result type cannot be represented in objective-c
```

Is this a bug with BrightFutures Objective-C support not added fully etc (#12 confuses me a bit) or simply because Objective-c do not have generics ?

And/Or do I simply declare the protocol wrongly ~ a here and now work around are very welcome!

**Comment by Thomvis** [comment]
I'm sorry for the confusion. BrightFutures currently has no support for Objective-C. It didn't make the cut for 1.0. 

I'm not sure how much demand there is and as long as I am not using BrightFutures in objective-c myself I am hesitant to add it. There is a PR that is already at 80%. Let me know if anyone wants to help.

**Comment by pellekrogholt** [comment]
OK thanks for clearing that out.

Concerning the `demand` for Objective-C support - its not clear to me either. My usecase might be a corner case but on the other hand a stack of `Dependency injection` with Typhoon some mock testing with OCMock might not be so exotic...

@Thomvis depending on the complexity of last 20% I would probably be interested to put in some work If you can assist me on what should be done and are willing to review it.

**Comment by Thomvis** [comment]
I've reopened the PR https://github.com/Thomvis/BrightFutures/pull/27. It needs to be rebased on the current master and some last features have to be ported. If every test in Swift has a succeeding equivalent in Objective-C, I'd say the porting is done.

However, in your example above it would still mean giving up the typesafety-ness of Swift, meaning that you cannot have a `Future<Bool>` (because Objc does not support generics) and instead will have a `BFFuture` that will have an `id`. Is that what you want?

**Comment by pellekrogholt** [comment]
@Thomvis exactly ...

i'll be back with some more in-depth questions soon'ish....

**Comment by pellekrogholt** [comment]
@Thomvis well this it not a perfect start but I'm facing a problem to do the rebase

i'm on the objc branch (forked) and do 

    $ git pull --rebase origin master    

but I have a problem to solve the `BrightFutures.xcodeproj/project.pbxproj` conflict.

could you by chance help me out with a rebase and push a refreshly rebased objc ?

**Comment by Thomvis** [comment]
Yes, I will do that. Probably later today or tomorrow. Would you mind if I rebase it off of 2.0-development? I'm hoping to release 2.0 in the coming weeks.

**Comment by pellekrogholt** [comment]
@Thomvis 2.0 that sounds sane for me

**Comment by Thomvis** [comment]
@pellekrogholt you could have a look at #54. As I describe in the PR, it is currently in a rough state (failing tests etc.) but the diff is nice and clean and should be easy to merge into 2.0-development once the work is done. I will not be able to work on it in the next week so feel free to give it a try and let me know what you think or what you want to do! Thanks.

**Comment by drosenstark** [comment]
For those of us who got here from the Internet, what's the current status of Objective-C with BrightFutures? Thanks!

**Comment by Thomvis** [comment]
There is currently no active effort to bring Objective-C support to BrightFutures, as far as I know. You are of course welcome to help by creating issues with specific compatibility issues or filing a PR.

**Comment by stowy** [comment]
Hi @Thomvis we are interested in using BrightFutures but have a mixed code base and would need objc support. Can you provide some insights as to what's needed to resurrect this effort?

**Comment by Thomvis** [comment]
Hi @stowy. #54 is pretty old and a lot has changed since. I don't see that ever getting merged. What I would do is just start using BF and write wrappers (somewhat similar to PromiseKit) as you need them. They don't even have to live in BF itself, they probably could be a separate framework.
